### PR TITLE
Result processor: gRPC pytest implementation

### LINKED ISSUE #1: Results Processing for HIL Tests
## High Level Overview
The HIL needs to be able to record data from a test, submit it, and compare it against a database of expected (passing) results. The goal is to create the HIL test database and results processing infrastructure. Each test state should contain a structure to record the data for that state and submit the results for the test.

## Data Types
Multiple data types need to be supported:
- boolean
- integer
- floating point
- arrays

## Database
The database should contain tags that store expected (passing) results for the test. Note: if no limits are specified, the value is assumed to be logged (and not compared)
- tag name
- data type
- expected value (if needed)
- upper limit (if needed)
- lower limit (if needed)

## Other Information
Have a look at the architecture built so far - there will be individual states that are executed during a test. Consider organizing the tags based on the state so each state can contain a structure to hold its recorded data. The database can be created as a package in the _hilmatic_ repo and should support storing the results for each test.

### PR DESCRIPTION
**Pull request by samparent97**
closes #1

### CODE REVIEW

File: results/proto/results.proto [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    string  value_str = 2;
+    int32   value_int = 3;
+    float   value_float = 4;
+    bool    value_bool = 5;
+  }
+}
+
+message SubmitTagResponse {
+  bool success = 1;
+  string error = 2;
```
Comment (samparent97) [inline results/proto/results.proto:27]: might be good to add these to all responses

File: results/server/demo/demo.py.j2 [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import pytest
+from ra_test import RATest
```
Comment (samparent97) [inline results/server/demo/demo.py.j2:2]: do we want to keep this file? if so lets update it with the newer jinja contents

File: results/server/demo/ra_test.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import sys
```
Comment (samparent97) [inline results/server/demo/ra_test.py:1]: Update this if we want to keep it

File: results/server/generated/results_pb2_grpc.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
+# Generated by the gRPC Python protocol compiler plugin. DO NOT EDIT!
+"""Client and server classes corresponding to protobuf-defined services."""
+import grpc
+
+from . import results_pb2 as results__pb2
```
Comment (samparent97) [inline results/server/generated/results_pb2_grpc.py:5]: This line is not autogen. If we dont care to fix it, let's add it to read me

File: results/server/main.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+tags_fp="./results/server/tags.yaml"
+tags_schema_fp="./results/server/schema/tags_schema.json"
+template_fp="./results/server/tm.py.j2"
+historic_tests_fp="../macfe-hil.github.io/historic_tests.json"
+reports_dir="../macfe-hil.github.io/reports"
+pages_repo_dir="../macfe-hil.github.io"
+pages_branch="main"
+git_username="macformularacing"
+git_email="macformulaelectric@gmail.com"
```
Comment (samparent97) [inline results/server/main.py:15]: should we read from the same config as the rest of the hil app?

File: results/server/main.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+def serve():
+    port = "8080"
```
Comment (samparent97) [inline results/server/main.py:18]: Lets change this to something that will not clash as often 
```python
port = "31763"
```

File: results/server/result_accumulator.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            existing_tests = json.load(file)
+
+        # Prepend the new test to the existing tests
+        existing_tests.insert(0, new_test)
+
+        # Save the updated list back to the JSON file
+        with open(self.historic_tests_fp, 'w') as file:
+            json.dump(existing_tests, file, indent=2)
+
+    def __push_to_github_pages(self, test_id) -> None:
```
Comment (samparent97) [inline results/server/result_accumulator.py:176]: Lets move this into a separate class.

File: results/server/server.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import grpc
+from concurrent import futures
+from proto import results_pb2_grpc
+from proto import results_pb2
+from result_accumulator import ResultAccumulator
+
```
Comment (samparent97) [inline results/server/server.py:6]: remove old file

File: results/server/tag_tunnel.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        error_count = self.ra.submit_error(request.error)
+        
+        return results_pb2.SubmitErrorResponse(error_count=error_count)
+
+    def SubmitTag(self, request, context):
+        # get the value of the oneof entry that is actually set
+        value = getattr(request, request.WhichOneof("data"))
+        is_passing, e = self.ra.submit_tag(request.tag, value)
+        if e != None:
+            return results_pb2.SubmitTagResponse(success=False, error=str(e), is_passing=is_passing)
```
Comment (samparent97) [inline results/server/tag_tunnel.py:65]: lets check the behaviour on this

File: results/server/tm.py.j2 [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+@pytest.fixture
+def tag_submissions():
+    return {{ tag_submissions }}
+
+@pytest.fixture
+def errors():
+    return {{ error_submissions }}
+
+# def test_overall(errors):
```
Comment (samparent97) [inline results/server/tm.py.j2:11]: Lets remove the commented out stuff

File: results/server/server.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        return results_pb2.SubmitTagReply(success=success)
+
+    def SubmitError(self, request, context):
+        """TODO: implement"""
+        self.ra.submit_error(request.error)
+        return results_pb2.SubmitErrorReply(success=True)
+
+    def CompleteTest(self, request, context):
+        self.ra.generate_and_run_tests()
+        return results_pb2.CompleteTestReply(success=True)
```
Comment (TylerStAmour) [inline results/server/server.py:28]: Shouldn't we get rid of this if we're going to use tag_tunnel.py?

**Comment by itshady** [comment]
lgtm. rely on @hsidd1 and @TylerStAmour for the more in-depth review tho

File: results/server/server.py (hunk 11 above)
Comment (samparent97) [reply results/server/server.py:28]: yes

File: results/server/main.py (hunk 6 above)
Comment (langei) [reply results/server/main.py:18]: I like the idea of adding this and all the other config parameters to the same config.yml as what is used in the hil app

File: results/server/server.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def serve():
+    port = "8080"
```
Comment (langei) [inline results/server/server.py:32]: update port here

File: results/server/demo/demo.py.j2 (hunk 2 above)
Comment (hsidd1) [reply results/server/demo/demo.py.j2:2]: We can update it when testing. Kinda just made this to run it without the server as a separate version. I find it helpful but we can remove it

File: results/server/main.py (hunk 5 above)
Comment (hsidd1) [reply results/server/main.py:15]: We can make yml within results for simpler organization

File: results/server/result_accumulator.py (hunk 7 above)
Comment (samparent97) [reply results/server/result_accumulator.py]: Resolved by @hsidd1's PR

File: results/server/demo/demo.py.j2 (hunk 2 above)
Comment (samparent97) [reply results/server/demo/demo.py.j2:2]: I'm fine with keeping the demo folder, just want to have it up to date

File: results/proto/results.proto (hunk 1 above)
Comment (samparent97) [reply results/proto/results.proto:27]: Created a gh issue for this #41

**Comment by samparent97** [comment]
This most recent commit just adds the requirements to generate the python grpc server code.

**Review by hsidd1 — APPROVED** [review]

**Review by TylerStAmour — APPROVED** [review]

File: results/client/client.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+import (
+	"context"
+	"github.com/google/uuid"
+	proto "github.com/macformula/hil/results/client/generated"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type ResultsClient struct {
```
Comment (langei) [inline results/client/client.go:12]: can we add some brief comments to our structs and public methods?

**Review by langei — APPROVED** [review]
Everything looks good

File: results/client/client.go (hunk 13 above)
Comment (samparent97) [reply results/client/client.go:12]: https://github.com/macformula/hil/issues/42

**Pull request merged**